tests/integration-tests/tests/networking/test_security_groups.py

```diff
@@ -160,56 +187,65 @@ def _check_connections_between_head_node_and_compute_nodes(cluster):
+
+    def _custom_security_group(count_of_sgs=1):
+        template = Template()
+        template.set_version("2010-09-09")
+        template.set_description("custom security group stack for testing additional_sg and vpc_security_group_id")
+        vpc_id = vpc_stack.cfn_outputs["VpcId"]
+        for sg_indx in range(count_of_sgs):
+            security_group = template.add_resource(
+                SecurityGroup(
+                    f"SecurityGroupResource{sg_indx}",
+                    GroupDescription="custom security group for testing additional_sg and vpc_security_group_id",
+                    VpcId=vpc_id,
+                )
             )
+            cidr_block_association_set = boto3.client("ec2").describe_vpcs(VpcIds=[vpc_id])["Vpcs"][0][
+                "CidrBlockAssociationSet"
+            ]
+            # Allow inbound connection within the VPC
+            for index, cidr_block_association in enumerate(cidr_block_association_set):
+                vpc_cidr = cidr_block_association["CidrBlock"]
+                template.add_resource(
+                    SecurityGroupIngress(
+                        f"SecurityGroupIngressResource{sg_indx}{index}",
+                        IpProtocol="-1",
+                        FromPort=0,
+                        ToPort=65535,
+                        CidrIp=vpc_cidr,
+                        GroupId=Ref(security_group),
+                    )
+                )
+            # Allow all inbound SSH connection
+            template.add_resource(
+                SecurityGroupIngress(
+                    f"SecurityGroupSSHIngressResource{sg_indx}",
+                    IpProtocol="tcp",
+                    FromPort=22,
+                    ToPort=22,
+                    GroupId=Ref(security_group),
+                    CidrIp="0.0.0.0/0",
+                )
+            )
+
+        stack = CfnStack(
+            name=generate_stack_name("integ-tests-custom-sg", request.config.getoption("stackname_suffix")),
+            region=region,
+            template=template.to_json(),
         )
-    # Allow all inbound SSH connection
-    template.add_resource(
-        SecurityGroupIngress(
-            "SecurityGroupSSHIngressResource",
-            IpProtocol="tcp",
-            FromPort=22,
-            ToPort=22,
-            GroupId=Ref(security_group),
-            CidrIp="0.0.0.0/0",
-        )
-    )
-    stack = CfnStack(
-        name=generate_stack_name("integ-tests-custom-sg", request.config.getoption("stackname_suffix")),
-        region=region,
-        template=template.to_json(),
-    )
-    cfn_stacks_factory.create_stack(stack)
+        stacks_factory.create_stack(stack)
+
+        return stack
 
-    yield stack
+    yield _custom_security_group
 
     if not request.config.getoption("no_delete"):
-        cfn_stacks_factory.delete_stack(stack.name, region)
+        stacks_factory.delete_all_stacks()
```

Comment:
do we really need a stack for each security group?
It's better to have the security group loop into the same CfnStack, they are logically related since used by the same tests and in this way it's easier to cleanup in case of failures/issues.

Reply:
I shouldnt have changed `delete_stack` to `delete_all_stacks ` function.

Initially I was creating two stacks which I thought was excessive! 
But I later changed to creating one stack with multiple Security groups and missed the cleanup function here.
So we are creating a single stack but with multiple securityGroups and will change the cleanup!